test/qa/test_allowlist_port.py

```diff
@@ -61,44 +63,233 @@ def test_allowlist_does_not_create_new_routes_when_adding_deleting_port_connecte
+
+        expected_message = allowlist.MSG_ALLOWLIST_PORT_RANGE_REMOVE_ERROR % (port.value.replace(":", " - "), port.protocol)
+        assert expected_message in str(ex)
+
+
+@pytest.mark.parametrize("tech,proto,obfuscated", lib.TECHNOLOGIES)
+@pytest.mark.parametrize("port", lib.PORTS_RANGE, ids=[f"{port.protocol}-{port.value}" for port in lib.PORTS_RANGE])
+def test_allowlist_port_range_twice_disconnected(tech, proto, obfuscated, port):
+    lib.set_technology_and_protocol(tech, proto, obfuscated)
+
+    with lib.Defer(sh.nordvpn.allowlist.remove.all):
+        for x in range(2):
+            allowlist.add_ports_to_allowlist([port])
+
         assert not firewall.is_active([port])
+
+
+@pytest.mark.parametrize("tech,proto,obfuscated", lib.TECHNOLOGIES)
+@pytest.mark.parametrize("port", lib.PORTS_RANGE, ids=[f"{port.protocol}-{port.value}" for port in lib.PORTS_RANGE])
+@pytest.mark.flaky(reruns=2, reruns_delay=90)
+@timeout_decorator.timeout(40)
+def test_allowlist_port_range_twice_connected(tech, proto, obfuscated, port):
+    lib.set_technology_and_protocol(tech, proto, obfuscated)
+
+    with lib.Defer(sh.nordvpn.allowlist.remove.all):
+        with lib.Defer(sh.nordvpn.disconnect):
+            sh.nordvpn.connect()
+
+            for x in range(2):
+                allowlist.add_ports_to_allowlist([port])
+
+            assert firewall.is_active([port])
+
+
+@pytest.mark.parametrize("tech,proto,obfuscated", lib.TECHNOLOGIES)
+@pytest.mark.parametrize("port", lib.PORTS_RANGE, ids=[f"{port.protocol}-{port.value}" for port in lib.PORTS_RANGE])
+def test_allowlist_port_range_when_port_from_range_already_allowlisted_disconnected(tech, proto, obfuscated, port):
+    port_range = port.value.split(":")
+    random_port_from_port_range = str(random.randint(int(port_range[0]), int(port_range[1])))
+
+    lib.set_technology_and_protocol(tech, proto, obfuscated)
+
+    with lib.Defer(sh.nordvpn.allowlist.remove.all):
+        already_allowlisted_port = lib.Port(random_port_from_port_range, port.protocol)
+        allowlist.add_ports_to_allowlist([already_allowlisted_port])
+        assert not firewall.is_active([already_allowlisted_port])
+
+        allowlist.add_ports_to_allowlist([port])
+        assert not firewall.is_active([port]) and not firewall.is_active([already_allowlisted_port])
+
+
+@pytest.mark.parametrize("tech,proto,obfuscated", lib.TECHNOLOGIES)
+@pytest.mark.parametrize("port", lib.PORTS_RANGE, ids=[f"{port.protocol}-{port.value}" for port in lib.PORTS_RANGE])
+@pytest.mark.flaky(reruns=2, reruns_delay=90)
+@timeout_decorator.timeout(40)
+def test_allowlist_port_range_when_port_from_range_already_allowlisted_connected(tech, proto, obfuscated, port):
+    port_range = port.value.split(":")
+    random_port_from_port_range = str(random.randint(int(port_range[0]), int(port_range[1])))
+
+    lib.set_technology_and_protocol(tech, proto, obfuscated)
+
+    with lib.Defer(sh.nordvpn.allowlist.remove.all):
```

Comment:
maybe we should think in the future tests, if is possible, to have a function executed after each test to reset to the defaults, and to login before starting the tests. In this way each test would not need to manually cleanup.